### ISSUE TITLE
The issue of Using OpenMP to run FFTW calculation in parallel

### ISSUE THREAD

**Issue by NanXiao** [FFTW/fftw3#81]
Hi all,

Greetings from me!  

My application uses `OpenMP` to spawn many threads, and every thread do a calculation. The simplified pseudo-code likes this:  

```
fftw_make_planner_thread_safe();

#pragma omp parallel
{
	/* Allocate resources */
        in = (double*) fftw_malloc(sizeof(double) * 2*N);
	out = (fftw_complex*) fftw_malloc(sizeof(fftw_complex) * (N + 2));
	plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
	plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);

	/* Do some assign values */
	......
	fftw_execute(plan_fft_forw);
	......
	fftw_execute(plan_fft_back);
	......

	/* Free resources */
	fftw_destroy_plan(plan_fft_back);
	fftw_destroy_plan(plan_fft_forw);
	fftw_free(out);
	fftw_free(in);
}
```

(1) According to the [manual](http://www.fftw.org/fftw3_doc/Installation-and-Supported-Hardware_002fSoftware.html#Installation-and-Supported-Hardware_002fSoftware), I pass `--enable-openmp` during configuration, and link the application with `-lfftw3_omp -lfftw3`. The build is OK, but once run it, the application will core dump. I check the [code](https://github.com/FFTW/fftw3/blob/811a672bdaedec4363272d9f7ed5fae56086aeb1/threads/openmp.c#L80), and find since `fftw_threads_register_planner_hooks()` is an empty function in `OpenMP` case, `fftw_make_planner_thread_safe()` doesn't take any effect.  

(2) I switch back to link application using `-lfftw3_threads -lfftw3`, but still use `OpenMP` directive in code. The application can run and without core dump until now (has lived for nearly half an hour).  

So my questions are like this:  
(1) For the pseudo-code, is there any potential issue in parallel environment?  
(2) Use  `OpenMP` directive while link with `-lfftw3_threads`, will there be some potential bugs?

Thanks very much in advance!

Best Regards
Nan Xiao

**Comment by matteo-frigo** [comment]
What you are trying to do is probably a bad idea.  The fftw planner is single-threaded, and the intended usage is that you call the planner from a single thread.  If you cannot that, the intended usage is that you grab whatever lock makes sense for your application, with the understanding that you may be holding that lock for a long time, which is bad.

We implemented the make_planner_thread_safe() API after years of people asking for it, since there is no alternative in certain environments (plugin-based applications).   Avoid it if you can.

**Comment by NanXiao** [comment]
@matteo-frigo : Thanks very much for your reply first! But I have a confusion. Per my understanding, there are 2 methods of using FFTW in multi-thread:   

(1) 
```
lock()
{
       plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
       plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);
}
unlock()
lock()
{
	fftw_destroy_plan(plan_fft_back);
	fftw_destroy_plan(plan_fft_forw);
}
unlock()
```
(2)  Like what I have posted:
```
fftw_make_planner_thread_safe();

#pragma omp parallel
......
```

If I understand right? What is the difference between these 2 methods? If I understand wrong, please help to elaborate it, thanks very much in advance!

**Comment by NanXiao** [comment]
@matteo-frigo @stevengj Could you give some feedback on my updated comment? Thanks very much in advance!

**Comment by stevengj** [comment]
The issue here is that `fftw_make_planner_thread_safe` is [implemented for POSIX threads](https://github.com/FFTW/fftw3/blob/fee0f966b2d3fae18019dd03a9bae338b4108d42/threads/threads.c#L482-L491) but [not for the OpenMP threads](https://github.com/FFTW/fftw3/blob/fee0f966b2d3fae18019dd03a9bae338b4108d42/threads/openmp.c#L78-L82) version.  It would be nice to have the latter.

**Comment by NanXiao** [comment]
@stevengj Thanks very much for your response! But I still have the following doubts:  

(1) Since `fftw_make_planner_thread_safe` isn't ready for [OpenMP threads](https://github.com/FFTW/fftw3/blob/fee0f966b2d3fae18019dd03a9bae338b4108d42/threads/openmp.c#L78-L82), if I use `-lfftw3_threads -lfftw3` to compile and link the program who uses `OpenMP` (like the following code), this should also be OK.  

(2)  For the following code:  

```
fftw_make_planner_thread_safe();

#pragma omp parallel
{
	/* Allocate resources */
        in = (double*) fftw_malloc(sizeof(double) * 2*N);
	out = (fftw_complex*) fftw_malloc(sizeof(fftw_complex) * (N + 2));
	plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
	plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);

	/* Do some assign values */
	......
	fftw_execute(plan_fft_forw);
	......
	fftw_execute(plan_fft_back);
	......

	/* Free resources */
	fftw_destroy_plan(plan_fft_back);
	fftw_destroy_plan(plan_fft_forw);
	fftw_free(out);
	fftw_free(in);
}
```

Maybe there are some performance issues, but it should not generate any errors when run in parallel environment.

Are my understandings right? Thanks very much in advance!

**Comment by stevengj** [comment]
(1) Yes, if you explicitly implement your own locks around creating/destroying plans, so that only one thread does this, then it should be fine.  `fftw_execute` itself is thread-safe.

(2) No, creating plans in parallel in multiple threads is not safe, as explained in http://fftw.org/doc/Thread-safety.html#Thread-safety, even if every thread creates its own plan.

**Comment by NanXiao** [comment]
@stevengj 
> (2) No, creating plans in parallel in multiple threads is not safe, as explained in [http://fftw.org/doc/Thread-safety.html#Thread-safety](http://fftw.org/doc/Thread-safety.html#Thread-safety), even if every thread creates its own plan.

But I have called `fftw_make_planner_thread_safe()` before:  

```
fftw_make_planner_thread_safe();

#pragma omp parallel
{
	plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
	plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);
        ......
	fftw_destroy_plan(plan_fft_back);
	fftw_destroy_plan(plan_fft_forw);
}
```

I think from code logic aspect, it is the same as following code:  
```
lock()
{
       plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
       plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);
}
unlock()
lock()
{
	fftw_destroy_plan(plan_fft_back);
	fftw_destroy_plan(plan_fft_forw);
}
unlock()
```

Except the performance may be downgrade. Thanks!

**Comment by stevengj** [comment]
make_planner_threadsafe does nothing for the OpenMP FFTW

**Comment by NanXiao** [comment]
@stevengj 
> make_planner_threadsafe does nothing for the OpenMP FFTW  

Yes, I have found it. So my current solution is still using `OpenMP` im my application:  

```
 fftw_make_planner_thread_safe();

#pragma omp parallel
{
	plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
	plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);
        ......
	fftw_destroy_plan(plan_fft_back);
	fftw_destroy_plan(plan_fft_forw);
}
```

But link `-lfftw3_threads -lfftw3` in linkage stage. Is it feasible solution? Thanks very much!

**Comment by davied9** [comment]
@NanXiao  have you considered this

```
fftw_make_planner_thread_safe();

#pragma omp parallel
{
	/* Allocate resources */
        in = (double*) fftw_malloc(sizeof(double) * 2*N);
	out = (fftw_complex*) fftw_malloc(sizeof(fftw_complex) * (N + 2));
        #pragma omp critical
        {
        	plan_fft_forw = fftw_plan_dft_r2c_1d(2*N, in, out,  FFTW_PATIENT);
        	plan_fft_back = fftw_plan_dft_c2r_1d(2*N, out, in,  FFTW_PATIENT);
        }

	/* Do some assign values */
	......
	fftw_execute(plan_fft_forw);
	......
	fftw_execute(plan_fft_back);
	......

	/* Free resources */
	fftw_free(out);
	fftw_free(in);
        #pragma omp critical
        {
		fftw_destroy_plan(plan_fft_back);
		fftw_destroy_plan(plan_fft_forw);
        }

}

```
